rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,987 @@
+
+As well as 'inside' and 'outside' we have four other flavors of dataflow nodes, all uniquely determined from the '
+inside' nodes:
+
+- 'entry' nodes are those inside with at least one dataflow input outside.
+- 'exit' nodes are those inside with at least one dataflow output outside, or which are considered 'external' in the
+  underlying dataflow graph (eg because they represent the result of the overall function).
+- 'input' nodes are those outside with at least one dataflow output inside.
+- 'output' nodes are those outside with at least one dataflow input inside.
+
+Index sets for these are cached with the sub-graph for performance.
+
+It is valid to have multiple entry nodes (we can bind a parameter for each). It may be valid to have multiple exit
+nodes (we can build a tuple of all such). It may be valid to have exit nodes which also contribute to other inside
+nodes (ie represent a 'tap' on an intermediate result).
+
+Sub-graphs are closed under:
+
+- Disjoint union.
+- Wrapping by a function with given attributes. This can be used to encode "Composite" functions, or to represent a
+  candidate kernel within a "Primitive" function. (By combining 'wrapping' with
+  'union' we can encode, eg, 'this sub-graph should be placed inside a primitive function which itself may have calls to
+  composite functions).
+- Substitution, which allows a sub-graph w.r.t. one dataflow graph to be transformed to match some other (typically
+  smaller) dataflow graph.
+
+Note that the Relay `PatternPartitoner` goes directly from `Expr` to partitioned `Expr` without stopping at any
+intermediate representation. It may be worth 'promoting' `SubGraph` out of Collage and into the standard `DFPattern`
+suite, we leave that to future work.
+
+### CandidatePartition
+
+A `CandidatePartition` pairs a `SubGraph` with a `Target`. All Collage search and measurement is in terms of candidate
+partitions.
+
+### PartitionRule
+
+A `PartitionRule` describes how to find a set of `CandidatePartitions`s for a `DataflowGraph`. This and `SubGraph`
+above are the core Collage datatypes. All partition rules implement the method:
+
+```
+virtual std::vector<CandidatePartition> AllCandidates(const DataflowGraph& dataflow_graph,
+                                                      const PartitionSpec& spec) const;
+```
+
+The candidates are allowed to overlap, and ultimately it is the job of the Collage searcher to find a selection of
+candidates which cover the whole Relay expression without overlap. There may be many thousands of candidates in flight
+during the Collage search.
+
+We currently have three distinct flavors of partitions:
+
+- For pattern-based BYOC integrations, individual `DFPattern`s are used to select the `"Composite"` functions to
+  offload, and those are grouped into a `"Primitive"` Relay function with a `"Compiler"` attribute.
+- For operator-based BYOC integrations, per-operator predicates indicate operators to offload, and again those are
```

Comment:
Can we do a simple utility and covert these into patterns as well in a very early layer in Collage ? -- so that we don't have to maintain yet another rule ?

Reply:
As soon as https://github.com/apache/tvm/pull/10759 is in I'll terminate all of this special casing (with extreme prejudice!).